src/VecSim/spaces/IP/IP_AVX512F_BW_VL_VNNI_INT8.h

```diff
@@ -0,0 +1,75 @@
+/*
+ *Copyright Redis Ltd. 2021 - present
+ *Licensed under your choice of the Redis Source Available License 2.0 (RSALv2) or
+ *the Server Side Public License v1 (SSPLv1).
+ */
+
+#include "VecSim/spaces/space_includes.h"
+
+static inline void InnerProductStep(int8_t *&pVect1, int8_t *&pVect2, __m512i &sum) {
+    __m256i temp_a = _mm256_loadu_epi8(pVect1);
+    __m512i va = _mm512_cvtepi8_epi16(temp_a);
+    pVect1 += 32;
+
+    __m256i temp_b = _mm256_loadu_epi8(pVect2);
+    __m512i vb = _mm512_cvtepi8_epi16(temp_b);
+    pVect2 += 32;
+
+    // _mm512_dpwssd_epi32(src, a, b)
+    // Multiply groups of 2 adjacent pairs of signed 16-bit integers in `a` with corresponding
+    // 16-bit integers in `b`, producing 2 intermediate signed 32-bit results. Sum these 2 results
+    // with the corresponding 32-bit integer in src, and store the packed 32-bit results in dst.
+    sum = _mm512_dpwssd_epi32(sum, va, vb);
```

Comment:
Why not use `_mm512_dpbusd_epi32`, avoiding loading only 32 elements and converting them to int16?

Reply:
nice catch!
Unfortunately, `_mm512_dpbusd_epi32` expects `a` to be unsigned 8-bit integers vecor, and `b` to be signed 8-bit vector
> Multiply groups of 4 adjacent pairs of unsigned 8-bit integers in a with corresponding signed 8-bit integers in b

So although indeed faster, it gives the wrong results 

```
Benchmark                                  Time             CPU   Iterations UserCounters...
--------------------------------------------------------------------------------------------
IP_dpwssd/Dim:512                       5.41 ns         5.41 ns    128425185 dist=-1.024k
IP_dpbusd/Dim:512                       3.26 ns         3.26 ns    218138693 dist=130.048k
IP_sanity_bm/Dim:512/iterations:1        490 ns          360 ns            1 dist=-1.024k
```

Reply:
Bummer... So there is no function for multiplying int8*int8 or uint8*uint8? Only an unsigned by a signed?

Reply:
nope
we might want to consider using `_mm512_unpackhi_epi8` and `_mm512_unpacklo_epi8` for `uint8_t`